File: trip-kit-booking/src/test/java/com/skedgo/tripkit/booking/InputFormTest.java
package com.skedgo.tripkit.booking;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Java6Assertions.assertThat;

@RunWith(RobolectricTestRunner.class)
public class InputFormTest {

    @Test
    public void NullInput() {
        InputForm inputForm = InputForm.from(null);
        assertThat(inputForm).isNull();
    }

    @Test
    public void EmptyInput() {
        InputForm inputForm = InputForm.from(new ArrayList<FormGroup>());
        assertThat(inputForm).isNotNull();
        List<FormField> formFieldItemList = inputForm.input();
        assertThat(formFieldItemList).hasSize(0);
    }

    @Test
    public void NonEmptyInput() {

        // Create first form
        List<FormGroup> list = new ArrayList<>();
        FormGroup item1 = new FormGroup();
        item1.setTitle("Requester Passenger");
        List<FormField> list1 = new ArrayList<>();
        for (int i = 0; i < 3; i++) {
            StringFormField item = new StringFormField();
            item.setTitle("First name");
            item.setType("string");
            item.setValue("value");
            item.setId("string");
            list1.add(item);
        }
        item1.setFields(list1);
        list.add(item1);

        // Create second form
        FormGroup item2 = new FormGroup();
        item1.setTitle("Requester Passenger");
        List<FormField> list2 = new ArrayList<>();
        for (int i = 0; i < 2; i++) {
            StepperFormField item = new StepperFormField();
            item.setTitle("Age");
            item.setType("stepper");
            item.setValue(10);
            item.setId("stepper");
            list2.add(item);
        }
        OptionFormField optionFormField = new OptionFormField();
        optionFormField.setType("option");
        list2.add(optionFormField);
        item2.setFields(list2);
        list.add(item2);

        // Testing
        InputForm inputForm = InputForm.from(list);
        assertThat(inputForm).isNotNull();
        List<FormField> formFieldItemList = inputForm.input();
        assertThat(formFieldItemList).hasSize(6);
        assertThat(formFieldItemList.get(0)).isInstanceOf(StringFormField.class);
        assertThat(formFieldItemList.get(1)).isInstanceOf(StringFormField.class);
        assertThat(formFieldItemList.get(2)).isInstanceOf(StringFormField.class);
        assertThat(formFieldItemList.get(4)).isInstanceOf(StepperFormField.class);
        assertThat(formFieldItemList.get(5)).isInstanceOf(OptionFormField.class);
    }

}

File: trip-kit-booking/src/test/java/com/skedgo/tripkit/booking/InputFormTest.kt
package com.skedgo.tripkit.booking

import com.skedgo.tripkit.booking.OptionFormField
import com.skedgo.tripkit.booking.StepperFormField
import com.skedgo.tripkit.booking.StringFormField
import org.assertj.core.api.Java6Assertions
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class InputFormTest {

    @Test
    fun nullInput() {
        val inputForm = InputForm.from(null)
        assertNull(inputForm)
    }

    @Test
    fun emptyInput() {
        val inputForm = InputForm.from(emptyList<FormGroup>())
        assertNotNull(inputForm)
        val formFieldItemList = inputForm?.input()
        assertEquals(0, formFieldItemList?.size)
    }

    @Test
    fun nonEmptyInput() {
        // Create first form group
        val list = mutableListOf<FormGroup>()
        val item1 = FormGroup().apply {
            title = "Requester Passenger"
            fields = mutableListOf<FormField>().apply {
                for (i in 0 until 3) {
                    add(StringFormField().apply {
                        title = "First name"
                        type = "string"
                        setValue("value")
                        id = "string"
                    })
                }
            }
        }
        list.add(item1)

        // Create second form group
        val item2 = FormGroup().apply {
            title = "Requester Passenger"
            fields = mutableListOf<FormField>().apply {
                for (i in 0 until 2) {
                    add(StepperFormField().apply {
                        title = "Age"
                        type = "stepper"
                        setValue(10)
                        id = "stepper"
                    })
                }
                add(OptionFormField().apply {
                    type = "option"
                })
            }
        }
        list.add(item2)

        // Testing
        val inputForm = InputForm.from(list)
        assertNotNull(inputForm)
        val formFieldItemList = inputForm?.input()
        assertEquals(6, formFieldItemList?.size)
        assert(formFieldItemList?.get(0) is StringFormField)
        assert(formFieldItemList?.get(1) is StringFormField)
        assert(formFieldItemList?.get(2) is StringFormField)
        assert(formFieldItemList?.get(3) is StepperFormField)
        assert(formFieldItemList?.get(4) is StepperFormField)
        assert(formFieldItemList?.get(5) is OptionFormField)
    }
}